File: src/main/java/java_spark/spark_project/TopNScore.java
package java_spark.spark_project;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;
import scala.actors.threadpool.Arrays;

import java.util.Iterator;

public class TopNScore {
    public static void main(String[] args) {
        SparkConf conf=new SparkConf().setAppName("TopN").setMaster("local");
        JavaSparkContext sc=new JavaSparkContext(conf);

        JavaRDD<String> inputRDD=sc.textFile("/home/hdc/data/score.txt");
//        System.out.println("!!!!!!!!!!!!!!!!!!");
        JavaPairRDD<String,Integer>scorePairs=inputRDD.mapToPair(
                new PairFunction<String, String, Integer>() {
            public Tuple2<String, Integer> call(String s) throws Exception {
                String [] str=s.split(" ");
//                System.out.println(str[0]+" "+str[1]);
                return new Tuple2<String, Integer>(str[0],Integer.parseInt(str[1]));
            }
        });

        JavaPairRDD<String,Iterable<Integer>>groupRDD=scorePairs.groupByKey();

        JavaPairRDD<String,Iterable<Integer>>scoreTopN=groupRDD.mapToPair(
                new PairFunction<Tuple2<String, Iterable<Integer>>, String, Iterable<Integer>>() {
            public Tuple2<String, Iterable<Integer>> call(Tuple2<String, Iterable<Integer>> s) throws Exception {

                Integer[]list=new Integer[100];
                Integer[]list1=new Integer[3];
                int i=0,j=0;
                String className=s._1;
                Iterator<Integer> scores=s._2.iterator();
                while(scores.hasNext()){
                    int temp=scores.next();
                    list[i]=temp;
                    i+=1;
                }
                int length=0;
                for (i=0;i<list.length;i++) {
                    if(list[i]!=null)
                        length++;
                }
                for(i=0;i<length;i++){
                    for(j=i+1;j<length;j++){
                        if(list[i]<list[j]){
                            int temp=list[i];
                            list[i]=list[j];
                            list[j]=temp;
                        }
                    }
                }
                for (i=0;i<3;i++) {
                    list1[i] = list[i];

                }
                return new Tuple2<String, Iterable<Integer>>(className, Arrays.asList(list1));
            }
        });
        scoreTopN.foreach(new VoidFunction<Tuple2<String, Iterable<Integer>>>() {
            public void call(Tuple2<String, Iterable<Integer>> s) throws Exception {
                System.out.println(s._1+":");
                Iterator<Integer>list=s._2.iterator();
                while(list.hasNext()){
                    System.out.println(list.next());
                }
                System.out.println("----------------------------------");
            }
        });
    }
}

File: src/main/scala/scala_spark/spark_project/TopNScore.scala
package scala_spark.spark_project

import org.apache.spark.{SparkConf, SparkContext}

import scala.collection.mutable.ArrayBuffer

object TopNScore {
  def main(args: Array[String]): Unit = {
    val conf=new SparkConf()
      .setAppName("TopNScore").setMaster("local")
    val sc=new SparkContext(conf)
    val lines=sc.textFile("/home/hdc/data/score.txt")
    val pairs=lines.map(x=>(x.split(" ")(0),x.split(" ")(1)))
    val groupRDD=pairs.groupByKey()
    groupRDD.foreach(x=>println(x._2.getClass.getSimpleName+"  "+x._2))
    val sortRDD=groupRDD.mapValues { y =>
      val list=y.toList
      val b1=new ArrayBuffer[Int]()
      for(i<-0 until(list.length)){
        b1+=list(i).toInt
      }
      for(i<-0 until b1.length-1){
        for(j<- i+1  until b1.length){
          if(b1(i)<b1(j)){
            var temp=b1(i)
            b1(i)=b1(j)
            b1(j)=temp
          }
        }
      }
      b1.take(3)
    }
    sortRDD.foreach(x=>println(x._1+" "+x._2))
  }
}
